File: libmonet/lib/palettes/core_palette.js
'use strict';
import { lstarFromInt } from '../utils/color_utils.js';
import { CAM16, HCT } from '../hct.js';
import { TonalPalette } from './tonal_palette.js';

export class CorePalette {
    constructor(argb) {
        const cam = CAM16.fromIntInViewingConditions(argb);
        var JSCompiler_inline_result = new HCT(cam.hue, cam.chroma, lstarFromInt(argb));
        const hue = JSCompiler_inline_result.hue;
        this.a1 = new TonalPalette(hue, Math.max(48, JSCompiler_inline_result.chroma));
        this.a2 = new TonalPalette(hue, 16);
        this.a3 = new TonalPalette(hue + 60, 24);
        this.n1 = new TonalPalette(hue, 4);
        this.n2 = new TonalPalette(hue, 8);
        this.error = new TonalPalette(25, 84);
    }
};

File: libmonet/lib/palettes/core_palette.dart
import 'package:libmonet/hct/cam16.dart';
import 'package:libmonet/hct/hct.dart';

import '../color.dart';
import '../math.dart';
import '../utils.dart';
import 'tonal_palette.dart';

class CorePalette {
  final TonalPalette a1;
  final TonalPalette a2;
  final TonalPalette a3;
  final TonalPalette n1;
  final TonalPalette n2;
  final TonalPalette error;

  CorePalette._(this.a1, this.a2, this.a3, this.n1, this.n2, this.error);

  factory CorePalette(ARGBColor argb) {
    final cam = CAM16.fromArgbInViewingConditions(argb);
    var seedHCT = new HCT(cam.hue, cam.chroma, lstarFromArgb(argb));
    final hue = seedHCT.hue;
    return CorePalette._(
      new TonalPalette.from(hue, Math.max(48, seedHCT.chroma).toDouble()),
      new TonalPalette.from(hue, 16),
      new TonalPalette.from(hue + 60, 24),
      new TonalPalette.from(hue, 4),
      new TonalPalette.from(hue, 8),
      new TonalPalette.from(25, 84),
    );
  }
}
